### PR TITLE
[WIP] Allowing Http/Https files as input 

### PR DESCRIPTION
**Pull request by kapilkd13**
To allow remote http/https files as input.  Refer Issue #466

### CODE REVIEW

**Comment by tetron** [comment]
I assume this is a work in progress, since it doesn't actually download files yet.

**Comment by kapilkd13** [comment]
Hi @tetron Yes, Currently its work in progress. Right now it solves the problem of validation error that was raised earlier. And the workflow error caused due to os.readlink on a web url.

File: cwltool/pathmapper.py [hunk 1]
```diff
@@ -139,6 +142,15 @@ def trim_listing(obj):
     if obj.get("location", "").startswith("file://") and "listing" in obj:
         del obj["listing"]
 
+# Download http Files
+def downloadHttpFile(httpurl):
+        r = requests.get(httpurl, stream=True)
```
Comment (tetron) [inline cwltool/pathmapper.py]: Can we use CacheControl to manage file caching?

File: cwltool/pathmapper.py [hunk 2]
```diff
@@ -139,6 +142,15 @@ def trim_listing(obj):
...
+        with NamedTemporaryFile(mode='wb', delete=False) as f:
+            for chunk in r.iter_content(chunk_size=1024):
```
Comment (tetron) [inline cwltool/pathmapper.py]: You'll get better throughput with a bigger chunk size (suggest at least 16k).

**Comment by kapilkd13** [comment]
Hi @mr-c Just to confirm, we need a conformance test for this feature right?

**Comment by mr-c** [comment]
@kapilkd13 Until https://github.com/common-workflow-language/common-workflow-language/issues/491 is fixed we can't add a conformance test for this as it is optional (though widely supported). So make it a test within `cwltool` only for now

File: cwltool/pathmapper.py [hunk 3]
```diff
@@ -139,6 +144,30 @@ def trim_listing(obj):
     if obj.get("location", "").startswith("file://") and "listing" in obj:
         del obj["listing"]
 
+# Download http Files
+def downloadHttpFile(httpurl):
+    cache_session = None
+    if "HOME" in os.environ:
```
Comment (mr-c) [inline cwltool/pathmapper.py]: Lets make this compliant with the [XDG Base Directory Specification](https://standards.freedesktop.org/basedir-spec/basedir-spec-0.6.html)'s `$XDG_CACHE_HOME` feature (if absent then the default is `$HOME/.config`) and if HOME isn't specified, then you can use [`os.path.expanduser('~')`](https://docs.python.org/3/library/os.path.html#os.path.expanduser) which means the bits with `TMP` and `/tmp` below aren't needed

Comment (kapilkd13) [inline cwltool/pathmapper.py]: I made these changes, i assumed you meant `$HOME/.cache` when `XDG_CACHE_HOME` is not present

**Comment by kapilkd13** [comment]
@mr-c  @tetron I think this feature is ready. Can you review this PR

File: cwltool/pathmapper.py [hunk 4]
```diff
@@ -139,6 +144,33 @@ def trim_listing(obj):
...
+    elif "HOME" in os.environ:
+        cache_session = CacheControl(
+            requests.Session(),
+            cache=FileCache(
+                os.path.join(os.environ["HOME"], ".cache", "cwltool")))
+    else:
+        cache_session = CacheControl(
+            requests.Session(),
+            cache=FileCache(
+                os.path.join(os.path.expanduser('~'), ".cache", "cwltool")))
```
Comment (mr-c) [inline cwltool/pathmapper.py]: Can you refactor this so the if/elif/else is just determining the directory to put `.cache` in and the call to CacheControl occurs afterwards?

**Comment by mr-c** [comment]
woo-hoo! Thank you @kapilkd13 !

**Pull request merged**